File: ruby/test/session_server_test.rb
# frozen_string_literal: true

require_relative "test_helper"
require "ed25519"

# Server-side session lifecycle tests. Mirrors
# `rust/crates/mpp/src/server/session.rs::tests`: open, voucher monotonicity,
# cap enforcement, min-delta, idempotent replay, top-up, metered delivery
# commit idempotency, and cooperative close.
class SessionServerTest < Minitest::Test
  include RubyMppTestHelpers

  S = ::Mpp::Protocol::Intents::Session

  # Deterministic Ed25519 session key from a fixed seed.
  def signer_key
    @signer_key ||= ::Ed25519::SigningKey.new("\x2a".b * 32)
  end

  def authorized_signer
    @authorized_signer ||= ::PayCore::Solana::Base58.encode(signer_key.verify_key.to_bytes)
  end

  def channel_id
    pubkey(9)
  end

  def make_server(min_voucher_delta: 0, max_cap: 10_000_000)
    config = ::Mpp::Server::Session::Config.new(
      operator: authorized_signer, recipient: pubkey(7),
      max_cap: max_cap, currency: "USDC", network: "localnet",
      min_voucher_delta: min_voucher_delta
    )
    ::Mpp::Server::Session.new(config: config, store: ::Mpp::MemoryChannelStore.new)
  end

  def open_server(server, deposit: 1_000_000)
    payload = S::OpenPayload.push(channel_id: channel_id, deposit: deposit, authorized_signer: authorized_signer, signature: "open_sig")
    server.process_open(payload)
  end

  # Build a signed voucher for `cumulative` on the test channel.
  def signed_voucher(cumulative, expires_at: S::DEFAULT_SESSION_EXPIRES_AT, channel: channel_id, key: signer_key)
    data = S::VoucherData.new(channel_id: channel, cumulative: cumulative.to_s, expires_at: expires_at)
    signature = ::PayCore::Solana::Base58.encode(key.sign(data.message_bytes))
    S::SignedVoucher.new(data: data, signature: signature)
  end

  # ── Challenge ──

  def test_build_challenge_request_clamps_cap_and_omits_push_only_modes
    server = make_server(max_cap: 5_000)
    request = server.build_challenge_request(1_000_000)

    assert_equal "5000", request.cap
    assert_empty request.modes
    assert_equal "localnet", request.network
    assert_equal "USDC", request.currency
  end

  def test_build_challenge_request_advertises_pull_strategy
    config = ::Mpp::Server::Session::Config.new(
      operator: authorized_signer, recipient: pubkey(7), network: "localnet",
      modes: [S::Mode::PUSH, S::Mode::PULL],
      pull_voucher_strategy: S::PullVoucherStrategy::OPERATED_VOUCHER
    )
    server = ::Mpp::Server::Session.new(config: config)
    request = server.build_challenge_request(1_000)

    assert_equal ["push", "pull"], request.modes
    assert_equal "operatedVoucher", request.pull_voucher_strategy
  end

  # ── Open ──

  def test_process_open_persists_channel_state
    server = make_server
    state = open_server(server, deposit: 500_000)

    assert_equal channel_id, state.channel_id
    assert_equal authorized_signer, state.authorized_signer
    assert_equal 500_000, state.deposit
    assert_equal 0, state.cumulative
    refute state.finalized
  end

  def test_process_open_rejects_zero_deposit
    server = make_server
    payload = S::OpenPayload.push(channel_id: channel_id, deposit: 0, authorized_signer: authorized_signer, signature: "s")

    assert_raises(::Mpp::VerificationError) { server.process_open(payload) }
  end

  def test_process_open_rejects_deposit_above_cap
    server = make_server(max_cap: 100)
    payload = S::OpenPayload.push(channel_id: channel_id, deposit: 1_000, authorized_signer: authorized_signer, signature: "s")

    assert_raises(::Mpp::VerificationError) { server.process_open(payload) }
  end

  def test_process_open_rejects_unsupported_mode
    server = make_server
    payload = S::OpenPayload.pull(token_account: pubkey(7), approved_amount: 1000, owner: pubkey(8), authorized_signer: authorized_signer, signature: "s")

    assert_raises(::Mpp::VerificationError) { server.process_open(payload) }
  end

  # ── Voucher ──

  def test_verify_voucher_advances_watermark
    server = make_server
    open_server(server)

    first = server.verify_voucher(S::VoucherPayload.new(voucher: signed_voucher(100)))
    second = server.verify_voucher(S::VoucherPayload.new(voucher: signed_voucher(250)))

    assert_equal 100, first
    assert_equal 250, second
  end

  def test_verify_voucher_rejects_non_increasing_cumulative
    server = make_server
    open_server(server)
    server.verify_voucher(S::VoucherPayload.new(voucher: signed_voucher(200)))

    assert_raises(::Mpp::VerificationError) do
      server.verify_voucher(S::VoucherPayload.new(voucher: signed_voucher(150)))
    end
  end

  def test_verify_voucher_idempotent_replay_returns_same_cumulative
    server = make_server
    open_server(server)
    voucher = signed_voucher(300)
    payload = S::VoucherPayload.new(voucher: voucher)

    first = server.verify_voucher(payload)
    replay = server.verify_voucher(payload)

    assert_equal 300, first
    assert_equal 300, replay
  end

  def test_verify_voucher_rejects_cumulative_above_deposit
    server = make_server
    open_server(server, deposit: 1000)

    assert_raises(::Mpp::VerificationError) do
      server.verify_voucher(S::VoucherPayload.new(voucher: signed_voucher(2000)))
    end
  end

  def test_verify_voucher_enforces_min_delta
    server = make_server(min_voucher_delta: 1000)
    open_server(server)
    server.verify_voucher(S::VoucherPayload.new(voucher: signed_voucher(1000)))

    error = assert_raises(::Mpp::VerificationError) do
      server.verify_voucher(S::VoucherPayload.new(voucher: signed_voucher(1500)))
    end
    assert_match(/below minimum/, error.message)

    # A voucher meeting the minimum delta is accepted.
    assert_equal 2000, server.verify_voucher(S::VoucherPayload.new(voucher: signed_voucher(2000)))
  end

  def test_verify_voucher_rejects_bad_signature
    server = make_server
    open_server(server)
    wrong_key = ::Ed25519::SigningKey.new("\x07".b * 32)
    voucher = signed_voucher(100, key: wrong_key)

    assert_raises(::Mpp::VerificationError) do
      server.verify_voucher(S::VoucherPayload.new(voucher: voucher))
    end
  end

  def test_verify_voucher_rejects_expired_voucher
    server = make_server
    open_server(server)
    voucher = signed_voucher(100, expires_at: 1)

    error = assert_raises(::Mpp::VerificationError) do
      server.verify_voucher(S::VoucherPayload.new(voucher: voucher))
    end
    assert_match(/expired/, error.message)
  end

  def test_verify_voucher_rejects_unknown_channel
    server = make_server

    assert_raises(::Mpp::VerificationError) do
      server.verify_voucher(S::VoucherPayload.new(voucher: signed_voucher(100)))
    end
  end

  # ── Top-up ──

  def test_process_topup_raises_deposit
    server = make_server
    open_server(server, deposit: 1000)

    state = server.process_topup(S::TopUpPayload.new(channel_id: channel_id, new_deposit: 5000, signature: "tx"))

    assert_equal 5000, state.deposit
  end

  def test_process_topup_rejects_non_increasing_deposit
    server = make_server
    open_server(server, deposit: 5000)

    assert_raises(::Mpp::VerificationError) do
      server.process_topup(S::TopUpPayload.new(channel_id: channel_id, new_deposit: 1000, signature: "tx"))
    end
  end

  def test_process_topup_rejects_above_cap
    server = make_server(max_cap: 2000)
    open_server(server, deposit: 1000)

    assert_raises(::Mpp::VerificationError) do
      server.process_topup(S::TopUpPayload.new(channel_id: channel_id, new_deposit: 3000, signature: "tx"))
    end
  end

  # ── Metered delivery commit ──

  def test_begin_delivery_then_commit_advances_watermark
    server = make_server
    open_server(server)

    directive = server.begin_delivery(::Mpp::Server::Session::DeliveryRequest.new(session_id: channel_id, amount: 100))

    assert_equal "100", directive.amount
    assert_equal 1, directive.sequence
    assert_equal "USDC", directive.currency

    receipt = server.process_commit(S::CommitPayload.new(delivery_id: directive.delivery_id, voucher: signed_voucher(100)))

    assert_equal "committed", receipt.status
    assert_equal "100", receipt.cumulative
    assert_equal "100", receipt.amount
  end

  def test_process_commit_is_idempotent_on_delivery_id
    server = make_server
    open_server(server)
    directive = server.begin_delivery(::Mpp::Server::Session::DeliveryRequest.new(session_id: channel_id, amount: 100))
    voucher = signed_voucher(100)
    payload = S::CommitPayload.new(delivery_id: directive.delivery_id, voucher: voucher)

    first = server.process_commit(payload)
    replay = server.process_commit(payload)

    assert_equal "committed", first.status
    assert_equal "replayed", replay.status
    assert_equal first.cumulative, replay.cumulative

    # The watermark did not double-advance.
    assert_equal 100, server.store.get_channel(channel_id).cumulative
  end

  def test_process_commit_rejects_different_voucher_for_committed_delivery
    server = make_server
    open_server(server)
    directive = server.begin_delivery(::Mpp::Server::Session::DeliveryRequest.new(session_id: channel_id, amount: 200))
    server.process_commit(S::CommitPayload.new(delivery_id: directive.delivery_id, voucher: signed_voucher(100)))

    # Re-commit the same delivery id with a different cumulative.
    assert_raises(::Mpp::VerificationError) do
      server.process_commit(S::CommitPayload.new(delivery_id: directive.delivery_id, voucher: signed_voucher(150)))
    end
  end

  def test_begin_delivery_rejects_amount_over_available_deposit
    server = make_server
    open_server(server, deposit: 100)

    assert_raises(::Mpp::VerificationError) do
      server.begin_delivery(::Mpp::Server::Session::DeliveryRequest.new(session_id: channel_id, amount: 200))
    end
  end

  def test_begin_delivery_rejects_duplicate_delivery_id
    server = make_server
    open_server(server)
    server.begin_delivery(::Mpp::Server::Session::DeliveryRequest.new(session_id: channel_id, amount: 100, delivery_id: "dup"))

    assert_raises(::Mpp::VerificationError) do
      server.begin_delivery(::Mpp::Server::Session::DeliveryRequest.new(session_id: channel_id, amount: 100, delivery_id: "dup"))
    end
  end

  def test_process_commit_unknown_delivery_raises
    server = make_server
    open_server(server)

    assert_raises(::Mpp::VerificationError) do
      server.process_commit(S::CommitPayload.new(delivery_id: "nope", voucher: signed_voucher(100)))
    end
  end

  # ── Close + finalize ──

  def test_process_close_applies_final_voucher_and_returns_finalize_params
    server = make_server
    open_server(server)
    server.verify_voucher(S::VoucherPayload.new(voucher: signed_voucher(100)))

    params = server.process_close(S::ClosePayload.new(channel_id: channel_id, voucher: signed_voucher(500)))

    assert_equal channel_id, params.channel_id
    assert_equal 500, params.settled
    assert_equal authorized_signer, params.authorized_signer
    assert_equal 32, params.distribution_hash.bytesize
    assert_equal pubkey(7), params.recipient
  end

  def test_process_close_without_voucher_keeps_watermark
    server = make_server
    open_server(server)
    server.verify_voucher(S::VoucherPayload.new(voucher: signed_voucher(250)))

    params = server.process_close(S::ClosePayload.new(channel_id: channel_id))

    assert_equal 250, params.settled
  end

  def test_voucher_rejected_after_close_requested
    server = make_server
    open_server(server)
    server.process_close(S::ClosePayload.new(channel_id: channel_id))

    assert_raises(::Mpp::VerificationError) do
      server.verify_voucher(S::VoucherPayload.new(voucher: signed_voucher(100)))
    end
  end

  def test_process_close_rejects_double_close
    server = make_server
    open_server(server)
    server.process_close(S::ClosePayload.new(channel_id: channel_id))

    assert_raises(::Mpp::VerificationError) do
      server.process_close(S::ClosePayload.new(channel_id: channel_id))
    end
  end

  def test_mark_finalized_blocks_further_vouchers
    server = make_server
    open_server(server)
    server.mark_finalized(channel_id)

    assert_raises(::Mpp::VerificationError) do
      server.verify_voucher(S::VoucherPayload.new(voucher: signed_voucher(100)))
    end
  end

  def test_payment_channel_open_params_validates_derived_channel
    config = ::Mpp::Server::Session::Config.new(
      operator: authorized_signer, recipient: pubkey(7), network: "localnet"
    )
    server = ::Mpp::Server::Session.new(config: config)
    mint = ::PayCore::Solana::Mints.resolve("USDC", "localnet")
    channel = ::Mpp::Program::PaymentChannels.channel_address(
      payer: pubkey(1), payee: pubkey(7), mint: mint, authorized_signer: authorized_signer, salt: 5
    )
    payload = S::OpenPayload.payment_channel(
      channel_id: channel, deposit: 1000, payer: pubkey(1), payee: pubkey(7),
      mint: mint, salt: 5, grace_period: 3600, authorized_signer: authorized_signer, signature: "s"
    )

    params = server.payment_channel_open_params(payload)

    assert_equal channel, params[:channel]
    assert_equal 5, params[:salt]
  end

  def test_payment_channel_open_params_rejects_mismatched_channel
    config = ::Mpp::Server::Session::Config.new(
      operator: authorized_signer, recipient: pubkey(7), network: "localnet"
    )
    server = ::Mpp::Server::Session.new(config: config)
    mint = ::PayCore::Solana::Mints.resolve("USDC", "localnet")
    payload = S::OpenPayload.payment_channel(
      channel_id: pubkey(2), deposit: 1000, payer: pubkey(1), payee: pubkey(7),
      mint: mint, salt: 5, grace_period: 3600, authorized_signer: authorized_signer, signature: "s"
    )

    assert_raises(::Mpp::VerificationError) { server.payment_channel_open_params(payload) }
  end
end

File: php/tests/Protocols/Mpp/Server/Session/SessionServerTest.php
<?php

declare(strict_types=1);

namespace PayKit\Tests;

use PHPUnit\Framework\TestCase;
use RuntimeException;
use PayKit\PayCore\Solana\Mints;
use PayKit\Protocols\Mpp\Core\PaymentChannels;
use PayKit\Protocols\Mpp\Intent\Session\ClosePayload;
use PayKit\Protocols\Mpp\Intent\Session\CommitPayload;
use PayKit\Protocols\Mpp\Intent\Session\CommitStatus;
use PayKit\Protocols\Mpp\Intent\Session\OpenPayload;
use PayKit\Protocols\Mpp\Intent\Session\SessionMode;
use PayKit\Protocols\Mpp\Intent\Session\SessionPullVoucherStrategy;
use PayKit\Protocols\Mpp\Intent\Session\SessionSplit;
use PayKit\Protocols\Mpp\Intent\Session\SignedVoucher;
use PayKit\Protocols\Mpp\Intent\Session\TopUpPayload;
use PayKit\Protocols\Mpp\Intent\Session\VoucherData;
use PayKit\Protocols\Mpp\Server\Session\DeliveryRequest;
use PayKit\Protocols\Mpp\Server\Session\MemoryChannelStore;
use PayKit\Protocols\Mpp\Server\Session\SessionConfig;
use PayKit\Protocols\Mpp\Server\Session\SessionServer;
use SolanaPhpSdk\Keypair\PublicKey;
use SolanaPhpSdk\Util\Base58;

final class SessionServerTest extends TestCase
{
    private const RECIPIENT = 'CXhrFZJLKqjzmP3sjYLcF4dTeXWKCy9e2SXXZ2Yo6MPY';

    /** @var array{0:string,1:string} [base58 pubkey, raw 64-byte secret] */
    private array $signerKey;

    private string $signer;

    private string $channel;

    protected function setUp(): void
    {
        $kp = sodium_crypto_sign_keypair();
        $pub = sodium_crypto_sign_publickey($kp);
        $this->signerKey = [Base58::encode($pub), sodium_crypto_sign_secretkey($kp)];
        $this->signer = $this->signerKey[0];
        $this->channel = PublicKey::fromBytes(str_repeat(chr(7), 32))->toBase58();
    }

    private function makeServer(int $minDelta = 0): SessionServer
    {
        return new SessionServer(
            new SessionConfig(
                operator: self::RECIPIENT,
                recipient: self::RECIPIENT,
                network: 'localnet',
                minVoucherDelta: $minDelta,
            ),
            new MemoryChannelStore(),
        );
    }

    private function voucher(int $cumulative, ?int $expiresAt = null): SignedVoucher
    {
        $expiresAt ??= time() + 3600;
        $data = new VoucherData($this->channel, (string) $cumulative, $expiresAt);
        $sig = sodium_crypto_sign_detached($data->messageBytes(), $this->signerKey[1]);
        return new SignedVoucher($data, Base58::encode($sig));
    }

    private function openPush(SessionServer $server, int $deposit): void
    {
        $server->processOpen(OpenPayload::push($this->channel, (string) $deposit, $this->signer, 'tx'));
    }

    public function testBuildChallengeRequestClampsCap(): void
    {
        $server = $this->makeServer();
        self::assertSame('10000000', $server->buildChallengeRequest(50_000_000)->cap);
        self::assertSame('5000000', $server->buildChallengeRequest(5_000_000)->cap);
    }

    public function testBuildChallengeRequestIncludesConfigFields(): void
    {
        $req = $this->makeServer()->buildChallengeRequest(1_000_000);
        self::assertSame(self::RECIPIENT, $req->operator);
        self::assertSame(self::RECIPIENT, $req->recipient);
        self::assertSame('USDC', $req->currency);
        self::assertSame(6, $req->decimals);
        self::assertSame('localnet', $req->network);
        self::assertSame([], $req->splits);
    }

    public function testProcessOpenStoresState(): void
    {
        $server = $this->makeServer();
        $state = $server->processOpen(OpenPayload::push($this->channel, '1000000', $this->signer, 'tx'));
        self::assertSame(1_000_000, $state->deposit);
        self::assertSame(0, $state->cumulative);
        self::assertFalse($state->finalized);
        self::assertSame($this->signer, $state->authorizedSigner);
    }

    public function testProcessOpenRejectsZeroDeposit(): void
    {
        $this->expectException(RuntimeException::class);
        $this->makeServer()->processOpen(OpenPayload::push($this->channel, '0', $this->signer, 'tx'));
    }

    public function testProcessOpenRejectsDepositOverCap(): void
    {
        $this->expectException(RuntimeException::class);
        $this->makeServer()->processOpen(OpenPayload::push($this->channel, '20000000', $this->signer, 'tx'));
    }

    public function testProcessOpenAtExactlyCapAccepted(): void
    {
        $server = $this->makeServer();
        $state = $server->processOpen(OpenPayload::push($this->channel, '10000000', $this->signer, 'tx'));
        self::assertSame(10_000_000, $state->deposit);
    }

    public function testProcessOpenRejectsUnadvertisedPullMode(): void
    {
        $server = $this->makeServer();
        $payload = OpenPayload::paymentChannel(
            SessionMode::Pull,
            $this->channel,
            '1000000',
            'payer',
            self::RECIPIENT,
            'mint',
            '1',
            900,
            $this->signer,
            'pending',
        );
        $this->expectExceptionMessage('not supported');
        $server->processOpen($payload);
    }

    public function testVerifyVoucherAdvancesWatermarkMonotonically(): void
    {
        $server = $this->makeServer();
        $this->openPush($server, 1_000_000);
        self::assertSame(100, $server->verifyVoucher($this->voucher(100)));
        self::assertSame(250, $server->verifyVoucher($this->voucher(250)));
    }

    public function testVerifyVoucherRejectsNonIncreasingCumulative(): void
    {
        $server = $this->makeServer();
        $this->openPush($server, 1_000_000);
        $server->verifyVoucher($this->voucher(250));
        $this->expectExceptionMessage('must exceed watermark');
        $server->verifyVoucher($this->voucher(100));
    }

    public function testVerifyVoucherIdempotentReplayReturnsSameWatermark(): void
    {
        $server = $this->makeServer();
        $this->openPush($server, 1_000_000);
        $v = $this->voucher(250);
        self::assertSame(250, $server->verifyVoucher($v));
        self::assertSame(250, $server->verifyVoucher($v));
    }

    public function testVerifyVoucherRejectsOverDeposit(): void
    {
        $server = $this->makeServer();
        $this->openPush($server, 500);
        $this->expectExceptionMessage('exceeds deposit');
        $server->verifyVoucher($this->voucher(600));
    }

    public function testVerifyVoucherEnforcesMinDelta(): void
    {
        $server = $this->makeServer(minDelta: 100);
        $this->openPush($server, 1_000_000);
        $server->verifyVoucher($this->voucher(100));
        $this->expectExceptionMessage('below minimum');
        $server->verifyVoucher($this->voucher(150));
    }

    public function testVerifyVoucherRejectsExpired(): void
    {
        $server = $this->makeServer();
        $this->openPush($server, 1_000_000);
        $this->expectExceptionMessage('expired');
        $server->verifyVoucher($this->voucher(100, time() - 1));
    }

    public function testVerifyVoucherRejectsBadSignature(): void
    {
        $server = $this->makeServer();
        $this->openPush($server, 1_000_000);
        $bad = new SignedVoucher(
            new VoucherData($this->channel, '100', time() + 3600),
            Base58::encode(str_repeat(chr(0), 64)),
        );
        $this->expectExceptionMessage('signature verification failed');
        $server->verifyVoucher($bad);
    }

    public function testVerifyVoucherUnknownChannel(): void
    {
        $this->expectExceptionMessage('not found');
        $this->makeServer()->verifyVoucher($this->voucher(100));
    }

    public function testMeteredDeliveryCommitAndIdempotentReplay(): void
    {
        $server = $this->makeServer();
        $this->openPush($server, 1_000);
        $directive = $server->beginDelivery(new DeliveryRequest($this->channel, 125));
        self::assertSame($this->channel, $directive->sessionId);
        self::assertSame('125', $directive->amount);
        self::assertSame(1, $directive->sequence);

        $voucher = $this->voucher(125);
        $payload = new CommitPayload($directive->deliveryId, $voucher);
        $receipt = $server->processCommit($payload);
        self::assertSame('125', $receipt->amount);
        self::assertSame('125', $receipt->cumulative);
        self::assertSame(CommitStatus::Committed, $receipt->status);

        $replay = $server->processCommit($payload);
        self::assertSame(CommitStatus::Replayed, $replay->status);
    }

    public function testBeginDeliveryRejectsOverDeposit(): void
    {
        $server = $this->makeServer();
        $this->openPush($server, 1_000);
        $server->beginDelivery(new DeliveryRequest($this->channel, 100));
        $this->expectExceptionMessage('exceeds available deposit');
        $server->beginDelivery(new DeliveryRequest($this->channel, 901));
    }

    public function testProcessCommitAcceptsPartialStreamUsage(): void
    {
        $server = $this->makeServer();
        $this->openPush($server, 1_000);
        $directive = $server->beginDelivery(new DeliveryRequest($this->channel, 125));
        $receipt = $server->processCommit(new CommitPayload($directive->deliveryId, $this->voucher(75)));
        self::assertSame('75', $receipt->amount);
        self::assertSame('75', $receipt->cumulative);
    }

    public function testProcessCommitRejectsOverReservedCumulative(): void
    {
        $server = $this->makeServer();
        $this->openPush($server, 1_000);
        $directive = $server->beginDelivery(new DeliveryRequest($this->channel, 125));
        $this->expectExceptionMessage('exceeds reserved amount');
        $server->processCommit(new CommitPayload($directive->deliveryId, $this->voucher(200)));
    }

    public function testProcessTopupRaisesDeposit(): void
    {
        $server = $this->makeServer();
        $this->openPush($server, 1_000_000);
        $state = $server->processTopup(new TopUpPayload($this->channel, '5000000', 'txsig'));
        self::assertSame(5_000_000, $state->deposit);
    }

    public function testProcessTopupRejectsNonIncreasingDeposit(): void
    {
        $server = $this->makeServer();
        $this->openPush($server, 1_000_000);
        $this->expectExceptionMessage('must exceed current deposit');
        $server->processTopup(new TopUpPayload($this->channel, '500000', 'txsig'));
    }

    public function testProcessCloseAppliesFinalVoucherAndBlocksFurtherVouchers(): void
    {
        $server = $this->makeServer();
        $this->openPush($server, 1_000_000);
        $server->verifyVoucher($this->voucher(100));
        $params = $server->processClose(new ClosePayload($this->channel, $this->voucher(500)));
        self::assertSame(500, $params->settled);
        self::assertSame(self::RECIPIENT, $params->recipient);
        self::assertSame(PaymentChannels::PROGRAM_ID, $params->programId);

        $this->expectExceptionMessage('close is pending');
        $server->verifyVoucher($this->voucher(600));
    }

    public function testFinalizeParamsCarriesDistributionHash(): void
    {
        $splitRecipient = PublicKey::fromBytes(str_repeat(chr(11), 32))->toBase58();
        $server = new SessionServer(
            new SessionConfig(
                operator: self::RECIPIENT,
                recipient: self::RECIPIENT,
                network: 'localnet',
                splits: [new SessionSplit($splitRecipient, 1000)],
            ),
            new MemoryChannelStore(),
        );
        $this->openPush($server, 1_000_000);
        $params = $server->finalizeParams($this->channel);
        self::assertSame(
            PaymentChannels::distributionHash([['recipient' => $splitRecipient, 'bps' => 1000]]),
            $params->distributionHash,
        );
        self::assertCount(1, $params->splits);
    }

    public function testMarkFinalizedBlocksVouchers(): void
    {
        $server = $this->makeServer();
        $this->openPush($server, 1_000_000);
        $server->markFinalized($this->channel);
        $this->expectExceptionMessage('finalized');
        $server->verifyVoucher($this->voucher(100));
    }

    public function testPaymentChannelOpenParamsValidatesChallengeFields(): void
    {
        $payer = PublicKey::fromBytes(str_repeat(chr(21), 32))->toBase58();
        $mint = Mints::resolve('USDC', 'localnet');
        self::assertIsString($mint);
        [$channel] = PaymentChannels::findChannelPda($payer, self::RECIPIENT, $mint, $this->signer, '77');

        $server = new SessionServer(
            new SessionConfig(
                operator: self::RECIPIENT,
                recipient: self::RECIPIENT,
                network: 'localnet',
                modes: [SessionMode::Pull],
                pullVoucherStrategy: SessionPullVoucherStrategy::ClientVoucher,
            ),
            new MemoryChannelStore(),
        );
        $payload = OpenPayload::paymentChannel(
            SessionMode::Pull,
            $channel->toBase58(),
            '1000000',
            $payer,
            self::RECIPIENT,
            $mint,
            '77',
            900,
            $this->signer,
            'pending',
        );
        $params = $server->paymentChannelOpenParams($payload);
        self::assertSame($payer, $params['payer']);
        self::assertSame($mint, $params['mint']);
        self::assertSame($channel->toBase58(), $params['channel']);

        // Wrong payee is rejected.
        $wrongPayee = OpenPayload::paymentChannel(
            SessionMode::Pull,
            $channel->toBase58(),
            '1000000',
            $payer,
            PublicKey::fromBytes(str_repeat(chr(22), 32))->toBase58(),
            $mint,
            '77',
            900,
            $this->signer,
            'pending',
        );
        $this->expectExceptionMessage('payee does not match');
        $server->paymentChannelOpenParams($wrongPayee);
    }

    public function testPaymentChannelOpenParamsRejectsMismatchedChannelPda(): void
    {
        $payer = PublicKey::fromBytes(str_repeat(chr(21), 32))->toBase58();
        $mint = Mints::resolve('USDC', 'localnet');
        self::assertIsString($mint);
        $server = new SessionServer(
            new SessionConfig(
                operator: self::RECIPIENT,
                recipient: self::RECIPIENT,
                network: 'localnet',
                modes: [SessionMode::Pull],
                pullVoucherStrategy: SessionPullVoucherStrategy::ClientVoucher,
            ),
            new MemoryChannelStore(),
        );
        $payload = OpenPayload::paymentChannel(
            SessionMode::Pull,
            PublicKey::fromBytes(str_repeat(chr(0xAB), 32))->toBase58(),
            '1000000',
            $payer,
            self::RECIPIENT,
            $mint,
            '77',
            900,
            $this->signer,
            'pending',
        );
        $this->expectExceptionMessage('channelId does not match');
        $server->paymentChannelOpenParams($payload);
    }
}
